### PR TITLE
Pin all unofficial github actions

### PR DESCRIPTION
**Pull request by Arkham**
### WHY are these changes introduced?

We use quite a lot of open source Github actions, which are inherently vulnerable to supply chain attacks.

### WHAT is this pull request doing?

- Pin all unofficial Github actions to a specific sha
- add a feature test to verify that all unofficial github actions are pinned

### How to test your changes?

- unpin any OSS github action
- run `cd packages/features`
- run `FEATURE=features/github-actions.feature pnpm test`
- you should see this test failing
```
1) Scenario: All non-official github actions are pinned # features/github-actions.feature:3
   ✔ When I look at the github actions we use # steps/github-actions.steps.ts:6
   ✖ Then I see all non-official actions being pinned # steps/github-actions.steps.ts:11
       AssertionError [ERR_ASSERTION]: The following unofficial github actions have not been pinned:

         - mxschmitt/action-tmate@v3

       Run the bin/pin-github-actions.js script, verify that the action is not doing anything malicious, then commit your changes.
           + expected - actual

           -1
           +0

           at World.<anonymous> (/Users/arkham/src/github.com/Shopify/cli/packages/features/steps/github-actions.steps.ts:28:10)
   ✔ After # steps/environment.steps.ts:38
   ```

### Measuring impact

How do we know this change was effective? Please choose one:

- [x] n/a - this doesn't need measurement, e.g. a linting rule or a bug-fix
- [ ] Existing analytics will cater for this addition
- [ ] PR includes analytics changes to measure impact

### Checklist

- [x] I've considered possible cross-platform impacts (Mac, Linux, Windows)
- [x] I've considered possible [documentation](https://shopify.dev) changes
- [x] I've made sure that any changes to `dev` or `deploy` have been reflected in the [internal flowchart](https://www.figma.com/file/7vqUp50u6dm48Zfb4JRRn8/CLI3-Internals).

### CODE REVIEW

**Comment by Arkham** [comment]
the `pin-github-action` script also reformatted all our yamls ✨

File: bin/pin-github-actions.js [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+import { exec } from 'child_process'
+
+export const execute = utils.promisify(exec);
+
+const __filename = fileURLToPath(import.meta.url);
+const __dirname = path.dirname(__filename);
+
+const args = process.argv;
+if (args.length !== 3) {
+  console.log(`Usage: bin/${path.basename(__filename)} <GITHUB_ACCESS_TOKEN>.\n\nThis script needs a Github access token to avoid hitting rate limits.\nYou can create one at https://github.com/settings/tokens.`)
```
Comment (pepicrft) [inline bin/pin-github-actions.js:16]: You can also detect if it's a Shopifolk and get it automatically by running `dev github print-auth`.

File: bin/pin-github-actions.js [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+  process.exit(1)
+}
+const githubAccessToken = args[2]
+
+async function doIt() {
```
Comment (pepicrft) [inline bin/pin-github-actions.js:21]: Do we need a function?

File: packages/features/steps/github-actions.steps.ts [hunk 3]
```diff
@@ -0,0 +1,29 @@
+import {exec} from '../lib/system'
+import {When, Then} from '@cucumber/cucumber'
+import * as path from 'pathe'
+import {strict as assert} from 'assert'
+
+When(/I look at the github actions we use/, async function () {
+  const {stdout} = await exec('grep', ['-Roh', 'uses: \\S*', '.'], {cwd: path.join(__dirname, '../../../.github')})
```
Comment (pepicrft) [inline packages/features/steps/github-actions.steps.ts:7]: Is it safe assuming that `grep` is in all the supported OSs?

**Review by pepicrft — APPROVED** [review]

File: packages/features/steps/github-actions.steps.ts (hunk 3 above)
Comment (Arkham) [reply packages/features/steps/github-actions.steps.ts:7]: yup, the windows environment on github actions has `/usr/bin/grep`, it's pretty neat

File: bin/pin-github-actions.js [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+  const githubYmls = glob.sync(`${__dirname}/../.github/{actions,workflows}/**/*.yml`)
+  const pinGithubAction = `${__dirname}/../node_modules/.bin/pin-github-action`
+  for (const githubYml of githubYmls) {
```
[the remark this answers is not in the document]
Comment (Arkham) [reply bin/pin-github-actions.js:24]: since it calls github apis I wanted to avoid DDOsing them :)

**Pull request merged**